creator/data_templates/mutations/template_version.py

```diff
@@ -0,0 +1,259 @@
+        required=True, description="The field definitions for this template"
+    )
+    data_template = graphene.ID(
+        required=True,
+        description="The data_template that this template_version belongs to",
+    )
+    studies = graphene.List(
+        graphene.ID,
+        description="The studies this template version should be assigned to",
+    )
+
+
+class UpdateTemplateVersionInput(graphene.InputObjectType):
+    """Parameters used when updating an existing template_version"""
+
+    description = graphene.String(
+        description="The description of the template_version"
+    )
+    field_definitions = graphene.JSONString(
+        description="The field definitions for this template"
+    )
+    studies = graphene.List(
+        graphene.ID,
+        description="The studies this template version should be assigned to",
+    )
+
+
+class CreateTemplateVersionMutation(graphene.Mutation):
+    """Creates a new template_version"""
+
+    class Arguments:
+        input = CreateTemplateVersionInput(
+            required=True,
+            description="Attributes for the new template_version",
+        )
+
+    template_version = graphene.Field(TemplateVersionNode)
+
+    def mutate(self, info, input):
+        """
+        Creates a new template_version.
+        """
+        user = info.context.user
+        if not user.has_perm("data_templates.add_datatemplate"):
+            raise GraphQLError("Not allowed")
+
+        # Check if data template exists
+        dt_id = input.pop("data_template")
+        model, dt_id = from_global_id(dt_id)
+
+        try:
+            dt = DataTemplate.objects.get(pk=dt_id)
+        except DataTemplate.DoesNotExist:
+            raise GraphQLError(f"DataTemplate {dt_id} does not exist")
+
+        # User may only add template versions to a template owned by their org
+        if not user.organizations.filter(pk=dt.organization.pk).exists():
+            raise GraphQLError(
+                "Not allowed - may only add template versions to templates "
+                "that are owned by the user's organization."
+            )
+
+        # Get study ids for later
+        study_ids = input.pop("studies", None)
+
+        # Create template version
+        with transaction.atomic():
+            template_version = TemplateVersion(**{k: input[k] for k in input})
+            template_version.creator = user
+            template_version.data_template = dt
+
+            # Validate and save template
+            template_version.clean()
+            template_version.save()
+
+            # Add studies to the template version if they exist
+            if study_ids:
+                # Ensure studies exist
+                studies = check_studies(study_ids)
+                template_version.studies.set(studies)
+                template_version.save()
+
+        return CreateTemplateVersionMutation(template_version=template_version)
+
+
+class UpdateTemplateVersionMutation(graphene.Mutation):
```

Comment:
I'm curious when would you update a template version vs creating a new template version with the changes

Reply:
I think template versions give us two things: 1) A way to have multiple studies/organizations collect the same set of concepts (participant diseases template) while allowing for some differences in how those concepts are defined (`Disease Category` vs `Diagnosis Type`) 2) A way to iterate on field definitions of a template and preserve history.

So I think then you would want to create a new `DataTemplate` when you want to represent a totally new concept/data type (participant phenotypes) and you would create a new `TemplateVersion` when you need to add or remove a field to/from the concept (e.g. add the `Observation Term` field to the template). Although, I guess there is nothing stopping someone from creating a new `TemplateVersion` with an entirely different set of columns compared to the current `TemplateVersion` and then that could change the semantics of the `DataTemplate` 🤔.

Reply:
thanks - this is a helpful description of new templates vs template versions. Although divergent column names feels like a pretty good reason to split off into a new data template no?

But I was actually asking about updating a templateVersion instead of creating a new templateVersion. 
Like, why would you update a templateVersion and not create a new templateVersion with your changes. 
I guess this lets you mutate the version until it is assigned to a study?

Reply:
| I guess this lets you mutate the version until it is assigned to a study?

Yep

Reply:
> Although divergent column names feels like a pretty good reason to split off into a new data template no?

Yes, _probably_, but I think we will just have to figure this out on a case by case basis. For example, if I add a column `description` to one of my templates, does that change the semantics of the template and therefore warrant a new data template? Probably not. But then if I add some columns and remove some columns to that same template, maybe that would warrant a new data template?

Reply:
This feels so much like a git workflow with tags and forks